### ISSUE TITLE
Unresolvable Asset from Future in dart:async/future.dart.

### ISSUE THREAD

**Issue by ganeshrvel** [Milad-Akarie/injectable#24]
I am trying to inject asynchronous SharedPreferences instance using injectable but the code is throwing me an error. The implementation:

```
import 'package:injectable/injectable.dart';
import 'package:shared_preferences/shared_preferences.dart';

@registerModule
abstract class SharedPreferencesDi {
  @lazySingleton
  Future<SharedPreferences> get sharedPreferences =>
      SharedPreferences.getInstance();
}
```

```
@injectable
@lazySingleton
class XXXDataSource extends SharedPreferencesDi {
  @override
  Future<SharedPreferences> sharedPreferences;

  XXXDataSource(this.sharedPreferences);
}
```

The error:

```
[INFO] Running build...
[SEVERE] injectable_generator:injectable_builder on lib/path/sources.dart:
Error running Instance of 'InjectableGenerator'
Unresolvable Asset from Future in dart:async/future.dart.
[WARNING] injectable_generator:injectable_config_builder on lib/di/di.dart:
Injectable Generator --------------------------------------------------

[XXXRepository] depends on [XXXDataSource] which is not injectable!

Did you forget to annotate the above classe(s) or their implementation with @injectable?
------------------------------------------------------------------------ 

[INFO] Running build completed, took 391ms

[INFO] Caching finalized dependency graph...
[INFO] Caching finalized dependency graph completed, took 103ms

[SEVERE] Failed after 497ms
```

**Comment by Milad-Akarie** [comment]
With the current setup, injectable will pre-await for SharedPreferences then register it as an instance
That will change in the future because There's now a register as async option in GetIT 4.0.0

also, you're not supposed the extend the register module and use it that way, think of it as a declaration space when you can register your third party dependencies 

but for the time being trey the following...
```dart

@registerModule
abstract class SharedPreferencesDi {
  @lazySingleton
  Future<SharedPreferences> get sharedPreferences =>
      SharedPreferences.getInstance();
}
// then inject it as an instance
// @injectable is redundant here @lazySingleton is enough 
@lazySingleton
class XXXDataSource {
  @override
 SharedPreferences sharedPreferences;
  XXXDataSource(this.sharedPreferences);
}
```

**Comment by ganeshrvel** [comment]
I have made the changes accordingly, but whenever there is a presence of the async option the program throws this error:

```
[ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: Invalid argument (Object of type XXXRepository is not registered inside GetIt.
 Did you forget to pass an instance name? 
(Did you accidentally do  GetIt sl=GetIt.instance(); instead of GetIt sl=GetIt.instance;)): AuthRepository
#0      GetIt.get (package:get_it/get_it.dart:76:7)
#1      GetIt.call (package:get_it/get_it.dart:87:12)
```

Code implementation:

```
@registerModule
abstract class SharedPreferencesDi {
  @lazySingleton
  Future<SharedPreferences> get sharedPreferences =>
      SharedPreferences.getInstance();
}
```

**Comment by Milad-Akarie** [comment]
can I see the generated code?

**Comment by ganeshrvel** [comment]
Yeah sure, here it is:

```
// GENERATED CODE - DO NOT MODIFY BY HAND

// **************************************************************************
// InjectableConfigGenerator
// **************************************************************************

import 'package:my_app/features/auth/data/data_sources/auth_local_data_source.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:my_app/common/local_storage/shared_preferences_di.dart';
import 'package:dio/dio.dart';
import 'package:my_app/common/api_client/dio_di.dart';
import 'package:my_app/common/api_client/api_client.dart';
import 'package:data_connection_checker/data_connection_checker.dart';
import 'package:my_app/common/network/network_info_di.dart';
import 'package:my_app/features/auth/data/api/auth_api.dart';
import 'package:my_app/features/auth/data/data_sources/auth_remote_data_source.dart';
import 'package:my_app/common/network/network_info.dart';
import 'package:my_app/features/auth/data/repository/auth_repository.dart';
import 'package:get_it/get_it.dart';

Future<void> $initGetIt(GetIt g, {String environment}) async {
  final sharedPreferencesDi = _$SharedPreferencesDi();
  final dioDi = _$DioDi();
  final networkInfoDi = _$NetworkInfoDi();
  g.registerLazySingleton<AuthLocalDataSource>(() => AuthLocalDataSource());
  final sharedPreferences = await sharedPreferencesDi.sharedPreferences;
  g.registerLazySingleton<SharedPreferences>(() => sharedPreferences);
  g.registerLazySingleton<Dio>(() => dioDi.dio);
  g.registerLazySingleton<ApiClient>(() => ApiClient(
        g<Dio>(),
      ));
  g.registerLazySingleton<DataConnectionChecker>(
      () => networkInfoDi.dataConnectionChecker);
  g.registerLazySingleton<AuthApi>(() => AuthApi(
        g<ApiClient>(),
      ));
  g.registerLazySingleton<AuthRemoteDataSource>(() => AuthRemoteDataSource(
        g<AuthApi>(),
      ));
  g.registerLazySingleton<NetworkInfo>(() => NetworkInfo(
        g<DataConnectionChecker>(),
      ));
  g.registerLazySingleton<AuthRepository>(() => AuthRepository(
        g<AuthLocalDataSource>(),
        g<AuthRemoteDataSource>(),
        g<NetworkInfo>(),
      ));
}

class _$SharedPreferencesDi extends SharedPreferencesDi {}

class _$DioDi extends DioDi {}

class _$NetworkInfoDi extends NetworkInfoDi {}
```

**Comment by Milad-Akarie** [comment]
The generated code looks perfect. 
Why isn't XXXDataSource registered though? 
Are you sure it's annotated with injectable/singleton?

**Comment by ganeshrvel** [comment]
@Milad-Akarie 

I have registered the injectable wherever it is required. Interestingly when I comment out the whole shared preferences di code the app works just fine. Once I un-comment it the app crashes with the error as mentioned in https://github.com/Milad-Akarie/injectable/issues/24#issuecomment-593902938

The link to the repo: https://github.com/ganeshrvel/open_cloud_encryptor/tree/from-the-scratch

clone the repo and checkout to from-the-scratch test the code.

Url to the https://github.com/ganeshrvel/open_cloud_encryptor/blob/from-the-scratch/lib/common/local_storage/shared_preferences_di.dart

**Comment by ganeshrvel** [comment]
@Milad-Akarie 
 
```
// The code works just fine when I comment async code. And when I substitute it with synchronous object the app runs just fine. I think the async code is messing it up

@registerModule
abstract class SharedPreferencesDi {
  /*
@lazySingleton
  Future<SharedPreferences> get sharedPreferences =>
      SharedPreferences.getInstance();
*/
}

```

**Comment by Milad-Akarie** [comment]
I think the problem is that you're not awaiting the initGetIt function 
Here's what you should do
```dart
@injectableInit
Future<void> setupDi() => $initGetIt(getIt);
```
then in your main
```dart
void main() async {
  await setupDi();
  ...
}
```

**Comment by cfregin** [comment]
Hey, Milad. Using version  injectable 0.2.3 and injectable_generator: 0.2.4  I do see a generator error:
`Future<SharedPreferences> get prefs => SharedPreferences.getInstance();`
translates to 
`final future<SharedPreferences> = await registerModule.prefs;`
(see the lowercase future?)

**Comment by cfregin** [comment]
```
class TestObj {
  static Future<TestObj> initialize() {
    return Future.value(TestObj());
  }
}

@registerModule
abstract class RegisterModule {
  @lazySingleton
  Dio get dio => Dio();

  Location get location => Location();

  Future<TestObj> get testGet => TestObj.initialize();
//Future<SharedPreferences> get prefs => SharedPreferences.getInstance();
}
```

for testing.

**Comment by ganeshrvel** [comment]
> injectable_generator: 0.2.4

@cfregin try 
`injectable_generator: 0.2.4`

**Comment by ganeshrvel** [comment]
> ```dart
> await setupDi();
> ```

@Milad-Akarie Yeah, I have made the change and now I get the error.

```
[INFO] Running build...
[SEVERE] injectable_generator:injectable_builder on lib/features/auth/data/data_sources/auth_local_data_source.dart:
Error running Instance of 'InjectableGenerator'
Unresolvable Asset from Future in dart:async/future.dart.
[WARNING] injectable_generator:injectable_config_builder on lib/common/di/di.dart:
Injectable Generator --------------------------------------------------

[AuthRepository] depends on [AuthLocalDataSource] which is not injectable!

Did you forget to annotate the above classe(s) or their implementation with @injectable?
```

and I double checked `AuthRepository` and `AuthLocalDataSource` are annotated using `@lazySingleton`

**Comment by Milad-Akarie** [comment]
@ganeshrvel  This should be fixed in v0.3.0
reOpen you still have issues.
